http-clients/java-nio-client/src/main/java/software/amazon/awssdk/http/nio/java/JavaHttpClientNioAsyncHttpClient.java

```diff
@@ -0,0 +1,271 @@
+    }
+
+    @Override
+    public String clientName() {
+        return CLIENT_NAME;
+    }
+
+
+    /**
+     * Builder that allows configuration of the Java NIO HTTP implementation. Use {@link #builder()} to configure
+     * and construct a Java Http Client.
+     */
+    public interface Builder extends SdkAsyncHttpClient.Builder<JavaHttpClientNioAsyncHttpClient.Builder> {
+
+        /**
+         * Maximum number of allowed concurrent requests.
+         *
+         * <p>
+         * If the maximum number of concurrent requests is exceeded they may queued in the HTTP Client and can
+         * cause latencies. If the client is overloaded enough such that the pending requests queue fills up,
+         * then subsequent requests may be rejected or time out.
+         * </p>
+         *
+         * @param maxConcurrency New value for max concurrency.
+         * @return This builder for method chaining.
+         */
+        Builder maxConcurrency(Integer maxConcurrency);
+
+
+        /**
+         * The amount of time to wait for a connection before an exeception is thrown.
+         *
+         * @param connectionTimeout timeout duration.
+         * @return This builder for method chaining.
+         */
+        Builder connectionTimeout(Duration connectionTimeout);
+
+
+        /**
+         * Sets the HTTP protocol to use (i.e. HTTP/1.1 or HTTP/2). Not all services support HTTP/2.
+         *
+         * @param protocol Protocol to use.
+         * @return This builder for method chaining.
+         */
+        Builder protocol(Protocol protocol);
+
+        /**
+         * Sets the SSL related parameters (e.g. Protocols, CipherSuites, ApplicationProtocols etc.) via SSLParameters
+         * object.
+         *
+         * @param sslParameter SSLParameters object.
+         * @return This builder for method chaining.
+         */
+        Builder configureSsl(SSLParameters sslParameter);
+
+        /**
+         * Sets the proxy related configurations (e.g. URI, port number etc.).
+         *
+         * @param proxyConfiguration ProxyConfiguration object.
+         * @return This builder for method chaining.
+         */
+        Builder proxyConfig(ProxyConfiguration proxyConfiguration);
+
+        /**
+         * Sets the amount of time to wait for a response before timeout.
+         *
+         * @param responseTimeout timeout duration.
+         * @return This builder for method chaining.
+         */
+        Builder responseTimeout(Duration responseTimeout);
+
+        /**
+         * Sets the numberOfThreads that the executor can hold in the threads pool.
+         *
+         * @param numberOfThreads an integer as the number of threads will be used in the executor.
+         * @return This builder for method chaining.
+         */
+        Builder numberOfThreads(Integer numberOfThreads);
+
+        /**
+         * Sets the executor in the HTTP Client as the executor created by users.
```

Comment:
I'm sorry I didn't get the point, by "this method" do you mean requestExecutor() or numberOfThreads(), and why shouldn't we close the executor if the customer uses it? These two methods will only be used while setting up the executor in the HttpClient, aren't they? Then why does it matter when the client is closed? Is it because we will have to check whether there are resources still held by the executor?

Reply:
The SDK as a policy doesn't manage resources created by the user. For example, if an operation takes an `InputStream`, we don't call `close()` on it; it is the responsbility of the caller. Basically, in most cases, the entity that created the resource is responsible for closing it.

This is similar to what happens in the Netty client. If the user provides a fully constructed `SdkEventLoopGroup`, it doesn't close it for them in the `close()` method.